ocaml/xapi/pool_periodic_update_sync.ml

```diff
@@ -0,0 +1,193 @@
+  in
+  let calc_delay =
+    Ptime.diff utc_next_schedule utc_now |> Ptime.Span.to_float_s
+  in
+  if Xapi_fist.disable_periodic_update_sync_sec_randomness () then
+    calc_delay
+  else
+    (* Make the minimum interval between 2 runs to 2 hours to avoid too close schedules *)
+    Float.max calc_delay update_sync_minimum_interval
+
+let update_sync_delay_for_next_schedule ~__context =
+  let frequency =
+    Db.Pool.get_update_sync_frequency ~__context
+      ~self:(Helpers.get_pool ~__context)
+  in
+  let day_configed =
+    Db.Pool.get_update_sync_day ~__context ~self:(Helpers.get_pool ~__context)
+  in
+  debug
+    "[PeriodicUpdateSync] update_sync_delay_for_next_schedule, frequency=%s, \
+     day_configed=%Ld"
+    (frequency_to_str ~frequency)
+    day_configed ;
+  let day_configed_int = Int64.to_int day_configed in
+  let utc_now = Ptime_clock.now () in
+  let tz_offset_s = Ptime_clock.current_tz_offset_s () |> Option.get in
+  let seconds_in_a_day = seconds_random_within_a_day () in
+  let utc_start_of_next_sched_day =
+    utc_start_of_next_scheduled_day ~utc_now ~tz_offset_s ~frequency
+      ~day_configed_int
+  in
+  let delay =
+    calc_delay ~utc_now ~utc_start_of_next_sched_day ~seconds_in_a_day
+  in
+  print_next_schedule ~delay ~utc_now ~tz_offset_s ;
+  delay
+
+let rec update_sync () =
+  Server_helpers.exec_with_new_task "periodic_update_sync" (fun __context ->
+      finally
+        (fun () ->
+          Helpers.call_api_functions ~__context (fun rpc session_id ->
+              try
+                ignore
+                  (Client.Pool.sync_updates ~rpc ~session_id
+                     ~self:(Helpers.get_pool ~__context)
+                     ~force:false ~token:"" ~token_id:""
+                  )
+              with e ->
+                let exc = Printexc.to_string e in
+                warn "Periodic update sync failed with exception %s" exc ;
+                let frequency =
+                  Db.Pool.get_update_sync_frequency ~__context
+                    ~self:(Helpers.get_pool ~__context)
+                in
+                let now =
+                  Ptime_clock.now ()
+                  |> Xapi_stdext_date.Date.of_ptime
+                  |> Xapi_stdext_date.Date.to_string
+                in
+                Xapi_alert.add ~msg:Api_messages.periodic_update_sync_failed
+                  ~cls:`Pool
+                  ~obj_uuid:
+                    (Db.Pool.get_uuid ~__context
+                       ~self:(Helpers.get_pool ~__context)
+                    )
+                  ~body:
+                    ("<body><message>Periodic update sync("
+                    ^ frequency_to_str ~frequency
+                    ^ ") failed.</message><exception>"
+                    ^ exc
+                    ^ "</exception><date>"
+                    ^ now
+                    ^ "</date></body>"
+                    )
+          )
+        )
```

Comment:
> I couldn't see anything in the design talking about this alert.

This alert is not from requirement, it is from a suggestion from Tina: since we do not retry on scheduled update sync failure, the updates may not be synced as user expected/configured, with this alert, user can choose to sync updates manually in case of failure if he or she wants.
I planed to implement it first, and will add into the design if it is agreed in the PR review and got merged.